src/utils/hibp.ts

```diff
@@ -0,0 +1,443 @@
+  }
+}
+/* c8 ignore stop */
+
+// TODO: Add unit test when changing this code:
+/* c8 ignore start */
+async function hibpApiFetch(path: string, options = {}) {
+  const url = `${HIBP_API_ROOT}${path}`;
+  const reqOptions = _addStandardOptions(options);
+  try {
+    return await _throttledFetch(url, reqOptions);
+  } catch (ex) {
+    console.error(ex);
+  }
+}
+/* c8 ignore stop */
+
+export type HibpGetBreachesResponse = Array<{
+  Name: string;
+  Title: string;
+  Domain: string;
+  AddedDate: ISO8601DateString;
+  BreachDate: ISO8601DateString;
+  DataClasses: Array<keyof HibpBreachDataTypes>;
+  Description: string;
+  LogoPath: string;
+  IsFabricated: boolean;
+  IsMalware: boolean;
+  IsRetired: boolean;
+  IsSensitive: boolean;
+  IsSpamList: boolean;
+  IsVerified: boolean;
+  ModifiedDate: ISO8601DateString;
+  PwnCount: number;
+}>;
+
+export async function fetchHibpBreaches(): Promise<HibpGetBreachesResponse> {
+  return (await hibpApiFetch("/breaches")) as HibpGetBreachesResponse;
+}
+
+// TODO: Add unit test when changing this code:
+/* c8 ignore start */
+async function kAnonReq(path: string, options = {}) {
+  // Construct HIBP url and standard headers
+  const url = `${HIBP_KANON_API_ROOT}${path}`;
+  options = {
+    headers: {
+      "Content-Type": "application/json",
+      Accept: "*/*",
+      "Hibp-Enterprise-Api-Key": HIBP_KANON_API_TOKEN,
+    },
+    ...options,
+  };
+
+  const reqOptions = _addStandardOptions(options);
+  try {
+    return await _throttledFetch(url, reqOptions);
+  } catch (ex) {
+    console.error(ex);
+  }
+}
+/* c8 ignore stop */
+
+/**
+ * Sanitize data classes
+ * ie. "Email Addresses" -> "email-addresses"
+ *
+ * @returns Array sanitized data classes array
+ */
+// TODO: Add unit test when changing this code:
+/* c8 ignore start */
+function formatDataClassesArray(dataClasses: Array<keyof HibpBreachDataTypes>) {
+  return dataClasses.map((dataClass) =>
+    dataClass
+      .toLowerCase()
+      // Replace non-alphanumeric characters, except for dashes, by dashes:
+      .replace(/[^-a-z0-9]/g, "-")
+      // Replace 2 or more consecutive dashes by a single dash:
+      .replace(/-{2,}/g, "-")
+      // Remove dashes at the start or end of the data class:
+      .replace(/(^-|-$)/g, ""),
+  ) as Array<HibpBreachDataTypes[keyof HibpBreachDataTypes]>;
+}
+/* c8 ignore stop */
+
+/**
+ * The type `HibpLikeDbBreach` is roughly the same as the data we receive from
+ * HIBP, except that we added a `FaviconUrl` and ID, that AddedDate,
+ * ModifiedDate and BreachDate are Date objects, and that a couple of fields are
+ * not available (because we do not store them in our database, at the time of
+ * writing).
+ *
+ * Presumably, our code used to be set up in such a way that it was supposed to
+ * work with objects that were either returned from the HIBP API, _or_ from our
+ * database - but the latter doesn't capitalise the properties.
+ */
+export type HibpLikeDbBreach = {
```

Comment:
Instead of basing this type on the HIBP response type, I explicitly based it on the types of the database row - after all, that's where we're getting these values from.

(That said, if you look at `knex-tables.d.ts`, you'll see that _those_ types are based on the HIBP response, if we don't do any further processing of them.)